ir/visitor.cpp

```diff
@@ -158,8 +155,121 @@ class Visitor::ChangeTracker {
      * if `start(@n)` has not been invoked.
      */
     const IR::Node *result(const IR::Node *n) const {
-        if (!visited.count(n)) return n;
-        return visited.at(n).result;
+        auto it = visited.find(n);
+        if (it == visited.end()) return n;
+        return it->second.result;
+    }
+
+    void visitOnce(const IR::Node *n) {
+        auto it = visited.find(n);
+        if (it == visited.end()) BUG("visitor state tracker corrupted");
+        it->second.visitOnce = true;
+    }
+
+    void visitAgain(const IR::Node *n) {
+        auto it = visited.find(n);
+        if (it == visited.end()) BUG("visitor state tracker corrupted");
+        it->second.visitOnce = false;
+    }
+};
+
+/** @class Visitor::Tracker
+ *  @brief Assists visitors in traversing the IR.
+
+ *  A Tracker object assists visitors traversing the IR by tracking each
+ *  node.  The `start` method begins tracking, and `finish` ends it.  The
+ *  `done` method determines whether the node has been visited.
+ */
+class Visitor::Tracker {
```

Comment:
Can we just use ChangeTracker instead of introducing a new class just for Inspector passes? ChangeTracker does track the result object that this doesn't -- I haven't carefully compared every function to see what other differences there are.

The downside though to reusing ChangeTracker is that it is heaver weight -- I think there are far more Inspector passes than Modifier/Transform passes, so perhaps it would be observable (e.g., memory use/GC impact)?

Reply:
It's more heavier. And also it expects that there is something as `result` doing few extra lookups.

The impact of `Inspectors` is quite large, so I'd keep the code separate, the difference is quite visible.